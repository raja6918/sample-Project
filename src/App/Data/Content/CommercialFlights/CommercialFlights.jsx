import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { union } from 'lodash';
import CommercialFlightstsForm from './CommercialFlightsForm';
import Notification from '../../../../components/Notification';
import Container from '../../../../components/DataContent/Container';
import AddHeader from '../../../../components/Headers/AddHeader';
import GenericTable from '../../../../components/GenericTable';
import withErrorHandler from '../../../../components/ErrorHandler/withErrorHandler';
import DataNotFound from '../../../../components/DataNotFound';
import TableMessage from '../../../../components/GenericTable/TableMessage';
import EditModeBar from '../../../../components/EditModeBar';
import Loading from '../../../../components/Loading';
import storage from '../../../../utils/storage';
import * as flightsService from '../../../../services/Data/commercialFlights';
import { getTimeReferenceConfig, checkEmptyFilter } from './helpers.js';
import { prepareSortPayload } from './../../../../components/GenericTable/helpers';
import { DEBUG_MODE } from './../../../../_shared/constants';
import DeleteDialog from '../../../../components/Dialog/DeleteDialog';
import withDeleteHandler from '../../../../components/DeleteHandler/withDeleteHandler';
import {
  orderBy,
  order,
  getHeaders,
  translationPath,
  TIME_REFERENCE,
  INITIAL_ITEMS,
  type,
} from './constants.js';
import * as stationsService from '../../../../services/Data/stations';
import AccessEnabler from '../../../../components/AccessEnabler';
import scopes from '../../../../constants/scopes';

export class CommercialFlights extends Component {
  state = {
    fetching: true,
    message: null,
    snackType: '',
    totalDataSize: 0,
    filteredDataSize: 0,
    flightCount: 0,
    isFilter: false,
    flightsData: [],
    isFormOpen: false,
    selectedItem: null,
  };

  commercialFlightsPayload = {
    startIndex: 0,
    endIndex: INITIAL_ITEMS,
    sort: prepareSortPayload({ field: orderBy }, order),
    filterCriteria: {},
  };

  componentDidMount() {
    const { openItemId } = this.props;

    Promise.all([
      flightsService.getCommercialFlights(
        openItemId,
        this.commercialFlightsPayload
      ),
      stationsService.getStations(openItemId),
      flightsService.getTags(openItemId),
    ])
      .then(([flights, stations, tags]) => {
        this.setState({
          flightsData: flights.data,
          totalDataSize: flights.totalDataSize,
          stations: stations.data,
          tags,
          fetching: false,
        });
      })
      .catch(err => {
        console.error(err);
      });
    this.setCommercialFlightCount();
  }

  handleEditItem = summaryItem => {
    const selectedItem = this.state.flightsData.find(
      flight => flight.id === summaryItem.id
    );

    this.setState(
      {
        selectedItem,
      },
      this.toggleForm
    );
  };

  handleSort = sortPayload => {
    const { openItemId } = this.props;

    this.commercialFlightsPayload = {
      ...this.commercialFlightsPayload,
      startIndex: 0,
      endIndex: INITIAL_ITEMS,
      sort: sortPayload,
    };

    flightsService
      .getCommercialFlights(openItemId, this.commercialFlightsPayload)
      .then(response => {
        const newFlightsData = response.data;
        this.setState({ flightsData: newFlightsData });
      });
  };

  handleFilter = filters => {
    const isEmpTyFilterObj = checkEmptyFilter(filters);
    const { openItemId } = this.props;
    this.commercialFlightsPayload = {
      ...this.commercialFlightsPayload,
      filterCriteria: filters,
      startIndex: 0,
      endIndex: INITIAL_ITEMS,
    };
    flightsService
      .getCommercialFlights(openItemId, this.commercialFlightsPayload)
      .then(response => {
        this.setState({
          flightsData: response.data,
          isFilter: isEmpTyFilterObj ? true : false,
          filteredDataSize: response.totalDataSize,
        });
      })
      .catch(err => {
        this.setState({
          flightsData: [],
          isFilter: isEmpTyFilterObj ? true : false,
          filteredDataSize: 0,
        });
        console.error(err);
      });
  };

  deleteCommercialFlight = flightAggregates => {
    const { id, flightDesignator } = flightAggregates;

    flightsService
      .deleteCommercialFlight(id, this.props.openItemId)
      .then(() => {
        const { flightsData, totalDataSize } = this.state;
        // this.setOperatingFlightCount().finally(() => {
        if (this.state.isFilter) {
          this.handleFilter(this.operatingFlightsPayload.filterCriteria);
        }
        this.setState({
          flightsData: flightsData.filter(item => item.id !== id),
          totalDataSize: totalDataSize - 1,
          message: this.props.t('SUCCESS.COMMERCIAL_FLIGHTS.REMOVE', {
            flightDesignator: flightDesignator,
          }),
          snackType: 'success',
        });
        // });
      })
      .catch(error => {
        error.errorType = 'Snackbar';
        this.props.handleDeleteError(error, type, flightDesignator);
      });
  };

  handleDelete = () => {
    this.deleteCommercialFlight(this.props.deleteDialogItem);
    this.props.closeDeleteDialog();
  };

  fetchData = nextItems => {
    const timerStart = Date.now();
    const { totalDataSize, isFilter, filteredDataSize } = this.state;
    const { openItemId } = this.props;
    const { endIndex: currentEndIndex } = this.commercialFlightsPayload;

    if (
      currentEndIndex === totalDataSize ||
      (totalDataSize > 0 && currentEndIndex > totalDataSize) ||
      (isFilter && currentEndIndex > filteredDataSize)
    )
      return;

    const nextStartIndex = currentEndIndex;
    let nextEndIndex = currentEndIndex + nextItems;

    if (nextEndIndex > totalDataSize) {
      nextEndIndex = totalDataSize;
    }

    this.commercialFlightsPayload = {
      ...this.commercialFlightsPayload,
      startIndex: nextStartIndex,
      endIndex: nextEndIndex,
    };

    flightsService
      .getCommercialFlights(openItemId, this.commercialFlightsPayload)
      .then(response => {
        if (DEBUG_MODE) {
          const elapsedTime = (Date.now() - timerStart) / 1000;
          // eslint-disable-next-line
          console.log('>>> Network time:', `${elapsedTime} seconds`);
        }

        const newFlightsData = [...this.state.flightsData, ...response.data];

        this.setState({ flightsData: newFlightsData });
      });
  };

  getScenarioDates = () => {
    const scenario = storage.getItem('openScenario');
    return {
      startDate: scenario ? scenario.startDate : '',
      endDate: scenario ? scenario.endDate : '',
    };
  };

  toggleForm = () => {
    const { isFormOpen, selectedItem } = this.state;
    if (isFormOpen && selectedItem) {
      this.setState({
        isFormOpen: !isFormOpen,
        selectedItem: null,
      });
    } else {
      this.setState({
        isFormOpen: !isFormOpen,
      });
    }
  };

  handleFlightForm = (flight, id) => {
    if (id === null) {
      this.addCommercialFlight(flight);
    } else {
      this.editCommercialFlight(flight, id);
    }
  };

  scrollToUpdatedItem = itemId => {
    window.setTimeout(() => {
      const element = document.getElementById(`tablerow${itemId}`);
      if (element) {
        element.scrollIntoView(true);
      }
    }, 200);
  };

  setCommercialFlightCount = () => {
    const { openItemId } = this.props;
    return flightsService
      .getCommercialFlightCount(openItemId)
      .then(flightCount => {
        this.setState({
          flightCount: flightCount.count,
        });
      })
      .catch(err => {
        console.error(err);
      });
  };

  addCommercialFlight = flight => {
    flightsService
      .addCommercialFlight(flight, this.props.openItemId)
      .then(r => {
        this.setCommercialFlightCount().finally(() => {
          if (this.state.isFilter) {
            this.handleFilter(this.commercialFlightsPayload.filterCriteria);
          }
          this.setState(prevState => ({
            flightsData: [r[0], ...prevState.flightsData],
            tags: union(prevState.tags, r[0].tags),
            totalDataSize: this.state.totalDataSize + 1,
            message: this.props.t('SUCCESS.COMMERCIAL_FLIGHTS.ADD', {
              flightNumber: r[0].flightNumber,
            }),
            snackType: 'success',
          }));
          this.toggleForm();
        });
      })
      .catch(error => {
        this.props.reportError({ error });
      });
  };

  editCommercialFlight = (flight, flightId) => {
    const flightsData = this.state.flightsData;
    flight.id = parseInt(flightId, 10);

    flightsService
      .editCommercialFlight(flight, this.props.openItemId)
      .then(r => {
        const editedFlight = r[0];

        const editedIndex = flightsData.findIndex(
          edditedFlight => edditedFlight.id === flightId
        );

        flightsData[editedIndex] = editedFlight;

        this.setCommercialFlightCount().finally(() => {
          if (this.state.isFilter) {
            this.handleFilter(this.commercialFlightsPayload.filterCriteria);
          }
          this.setState(
            prevState => ({
              flightsData,
              tags: union(prevState.tags, r[0].tags),
              message: this.props.t('SUCCESS.COMMERCIAL_FLIGHTS.EDIT', {
                flightNumber: r[0] ? r[0].flightNumber : '',
              }),
              snackType: 'success',
            }),
            () => this.scrollToUpdatedItem(flightId)
          );
          this.toggleForm();
        });
      })
      .catch(error => {
        this.props.reportError({ error });
      });
  };

  onClearSnackBar = () => {
    this.setState({
      message: null,
      snackBarType: '',
    });
  };

  render() {
    const {
      fetching,
      totalDataSize,
      flightsData,
      isFormOpen,
      stations,
      message,
      snackType,
      selectedItem,
      tags,
      isFilter,
      filteredDataSize,
      flightCount,
    } = this.state;
    const {
      t,
      editMode,
      openDeleteDialog,
      deleteDialogIsOpen,
      deleteDialogItem,
      exitDeleteDialog,
      closeDeleteDialog,
      readOnly,
      ...rest
    } = this.props;
    const dataScopes = editMode
      ? scopes.dataCardPage.dataTemplate
      : scopes.dataCardPage.dataScenario;
    const [timeReferenceText] = getTimeReferenceConfig(TIME_REFERENCE, t);

    if (fetching) {
      return <Loading />;
    } else {
      return (
        <React.Fragment>
          <AddHeader
            t={t}
            name={t('DATA.commercialFlights.name')}
            className="tm-scenario_data_commercial_flights__create-btn"
            onClick={this.toggleForm}
            disableAdd
            scopes={dataScopes}
            {...rest}
          />
          <Container>
            {editMode && <EditModeBar title={t('TEMPLATES.editMode')} />}
            <TableMessage>
              <span className="bold">
                {`Time reference: ${timeReferenceText}`}
              </span>
            </TableMessage>
            {flightsData.length || isFilter ? (
              <AccessEnabler
                scopes={dataScopes}
                disableComponent
                render={props => (
                  <GenericTable
                    headers={getHeaders(t, this.getScenarioDates().startDate)}
                    data={flightsData}
                    totalDataSize={totalDataSize}
                    orderBy={orderBy}
                    order={order}
                    handleEditItem={this.handleEditItem}
                    handleDeleteItem={openDeleteDialog}
                    handleFilter={this.handleFilter}
                    handleFetchData={this.fetchData}
                    handleSort={this.handleSort}
                    isFilter={isFilter}
                    filteredDataSize={filteredDataSize}
                    name={t(`${translationPath}.footerName`, {
                      flightCount,
                    })}
                    frontEndFilterdisabled={true}
                    t={t}
                    editMode={editMode}
                    handleDisableEdit={() => props.disableComponent}
                    handleDisableDelete={() => props.disableComponent}
                  />
                )}
              />
            ) : (
              <DataNotFound
                text={t('GLOBAL.dataNotFound.message', {
                  data: t(`${translationPath}.name`).toLowerCase(),
                })}
              />
            )}
            <AccessEnabler
              scopes={dataScopes}
              disableComponent
              render={props => (
                <CommercialFlightstsForm
                  isOpen={isFormOpen}
                  t={t}
                  handleOk={this.handleFlightForm}
                  stations={stations}
                  handleCancel={this.toggleForm}
                  onClose={this.toggleForm}
                  selectedItem={selectedItem}
                  tagsSource={tags}
                  enableReadOnly={props.disableComponent}
                  readOnly={readOnly}
                />
              )}
            />
            <Notification
              message={message}
              clear={this.onClearSnackBar}
              type={snackType}
            />
            <DeleteDialog
              open={deleteDialogIsOpen}
              handleOk={this.handleDelete}
              onExited={exitDeleteDialog}
              handleCancel={closeDeleteDialog}
              modalCommand={true}
              onClose={closeDeleteDialog}
              okText={t('GLOBAL.form.delete')}
              title={t('DIALOG.DELETE_DATA.title', {
                type: t(type).toLowerCase(),
                name: deleteDialogItem.flightDesignator || ' ',
              })}
              bodyText={t('DIALOG.DELETE_DATA.body', {
                type: t(type).toLowerCase(),
                name: deleteDialogItem.flightDesignator || ' ',
              })}
              t={t}
            />
          </Container>
        </React.Fragment>
      );
    }
  }
}

CommercialFlights.propTypes = {
  t: PropTypes.func.isRequired,
  openItemId: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  readOnly: PropTypes.bool.isRequired,
  reportError: PropTypes.func.isRequired,
  editMode: PropTypes.bool.isRequired,
  openDeleteDialog: PropTypes.func.isRequired,
  handleDeleteError: PropTypes.func.isRequired,
  deleteDialogItem: PropTypes.shape().isRequired,
  deleteDialogIsOpen: PropTypes.bool.isRequired,
  closeDeleteDialog: PropTypes.func.isRequired,
  exitDeleteDialog: PropTypes.func.isRequired,
};

export default withErrorHandler(withDeleteHandler(CommercialFlights)(type));
