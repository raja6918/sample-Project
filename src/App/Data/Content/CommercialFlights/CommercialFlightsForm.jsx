import React, { Component } from 'react';
import PropTypes from 'prop-types';
import PerfectScrollbar from 'react-perfect-scrollbar';
import pick from 'lodash/pick';
import 'react-perfect-scrollbar/dist/css/styles.css';

import Grid from '@material-ui/core/Grid';

import Form from '../../../../components/FormDrawer/Form';
import FormHeader from '../../../../components/FormDrawer/FormHeader';
import FormBody from '../../../../components/FormDrawer/FormBody';
import Calendar from '../../../../components/Calendar/Calendar';
import AutocompleteChips from '../../../../components/AutocompleteChips/AutocompleteChips';

import TextField from '@material-ui/core/TextField';
import Autocomplete from '@material-ui/lab/Autocomplete';
import ListboxComponent from '../../../../components/ReactWindow/ListboxComponent';

import { SelectInput, FormControl } from '../Accommodations/FormComponents';
import InputLabel from '@material-ui/core/InputLabel';
import storage from '../../../../utils/storage';
import { getStringHours } from '../../../../_shared/helpers';
import {
  perfectScrollConfig,
  getFormHeading,
  getFormOkButton,
  getFormCancelButton,
} from '../../../../utils/common';

import {
  CARRY_IN_DAYS,
  CARRY_OUT_DAYS,
} from '../../../Pairings/components/Timeline/constants';

import {
  GeneralTitle,
  InputForm,
  TimePickers,
  CustomText,
  SingleItem,
} from './FormComponents';

import {
  getDefaultEntity,
  getDefaultErrors,
  mapEntityToState,
  hasError,
  wereAllFieldsFilled,
  getCustomPredicate,
  prepareStations,
  findTerminalsByStationCode,
  sanitizeFlight,
  getStationName,
} from './helpers';

import {
  COMMERCIAL_FLIGHTS_FIELDS,
  COMMERCIAL_FLIGHTS_GENERAL,
  COMMERCIAL_FLIGHTS_FORM,
  COMMERCIAL_FLIGHTS_ERRORS,
  COMMERCIAL_FLIGHTS_INFO,
} from './constants';

class CommercialFlightsForm extends Component {
  state = {
    ...getDefaultEntity(),
    isFormDirty: false,
    errors: getDefaultErrors(),
  };

  startDate = null;
  endDate = null;

  componentDidMount() {
    const currentScenario = storage.getItem('openScenario');
    this.startDate = currentScenario.startDate;
    this.endDate = currentScenario.endDate;
  }

  componentWillReceiveProps(nextProps) {
    const { selectedItem, isOpen } = nextProps;
    if (selectedItem) selectedItem.tags = [];
    if (!isOpen) {
      const entity = selectedItem
        ? mapEntityToState(selectedItem)
        : getDefaultEntity();

      this.setState({
        ...entity,
        isFormDirty: false,
        errors: getDefaultErrors(),
      });
    }
  }

  handleChange = (e, shouldValidateForm = true) => {
    this.onChange(e.target.name, e.target.value, shouldValidateForm);
  };

  handleSelectChange = (option, stateProperty) => {
    const value = option ? option.value : '';

    let newState = {
      [stateProperty]: value,
    };

    switch (stateProperty) {
      case 'departureStationCode':
        newState = {
          ...newState,
          passengerTerminalDeparture: null,
        };
        break;
      case 'arrivalStationCode':
        newState = {
          ...newState,
          passengerTerminalArrival: null,
        };

        break;
      default:
        break;
    }

    this.setState(newState, this.checkFormIsReady);
  };

  handleInputChange = e => {
    const errors = this.updateErrors(e);
    let { name, value } = e.target;
    value = [
      'aircraftType',
      'aircraftConfigurationVersion',
      'airlineCode',
    ].includes(name)
      ? value.trim().toUpperCase()
      : value;

    let newState = {
      errors,
      [name]: value,
      onwardOffsetIsDisabled: false,
    };

    if (name === 'onwardFlightDesignator') {
      if (!value) {
        newState = {
          ...newState,
          onwardFlightDayOffset: '',
          onwardOffsetIsDisabled: true,
        };
      }
    }

    this.setState(newState, this.checkFormIsReady);
  };

  onChange = (fieldName, value, shouldValidForm = true) => {
    this.setState(
      {
        [fieldName]: value,
      },
      () => {
        if (shouldValidForm) this.checkFormIsReady();
      }
    );
  };

  checkFormIsReady = () => {
    const { isFormDirty } = this.state;
    const _hasError = hasError(this.state.errors);
    const allFieldsWereFilled = wereAllFieldsFilled(this.state);

    if (allFieldsWereFilled && !_hasError) {
      if (!isFormDirty) {
        this.setState({
          isFormDirty: true,
        });
      }
    } else {
      if (isFormDirty) {
        this.setState({
          isFormDirty: false,
        });
      }
    }
  };

  updateErrors = e => {
    const event = e.target || e.srcElement;
    const fieldName = event.name;
    const isValid = getCustomPredicate(fieldName);
    const isError = event.value !== '' && !isValid(event.value, this.state);
    return {
      ...this.state.errors,
      [fieldName]: isError,
    };
  };

  isReadyToSubmit = () => {
    return !this.state.isFormDirty;
  };

  handleSubmit = () => {
    const entity = sanitizeFlight(pick(this.state, COMMERCIAL_FLIGHTS_FIELDS));
    const id = this.props.selectedItem ? this.props.selectedItem.id : null;

    entity.startTime = getStringHours(entity.startTime);
    entity.endTime = getStringHours(entity.endTime);

    this.props.handleOk(entity, id);
  };

  handleChangePicker = name => dateTime => {
    this.setState({ [name]: dateTime }, this.checkFormIsReady);
  };

  handleCrewComplements = crewComposition => {
    this.setState(
      { crewComposition },
      this.props.handleCrewComplements(crewComposition)
    );
    if (crewComposition !== this.state.crewComposition) {
      this.checkFormIsReady();
    }
  };

  handleCalendarUpdates = date => {
    const startDates = [...this.state.startDates];
    const index = startDates.indexOf(date);

    if (index > -1) {
      startDates.splice(index, 1);
    } else {
      startDates.push(date);
    }
    this.setState({ startDates }, this.checkFormIsReady);
  };

  handleTagsUpdate = tags => {
    this.setState({ tags }, this.checkFormIsReady);
  };

  handleRemoveTag = tag => {
    const { tags } = this.state;
    const index = tags.indexOf(tag);

    if (index !== -1) {
      tags.splice(index, 1);
    }
    this.setState({ tags });
  };

  getGeneralSection = () => {
    const { t, stations, readOnly, enableReadOnly } = this.props;

    const {
      errors,
      airlineCode,
      flightNumber,
      operationalSuffix,
      departureStationCode,
      arrivalStationCode,
      passengerTerminalDeparture,
      passengerTerminalArrival,
      startTime,
      endTime,
      aircraftType,
      aircraftConfigurationVersion,
      startDates,
    } = this.state;
    const sectionPath = COMMERCIAL_FLIGHTS_GENERAL;
    const readOnlyStatus = readOnly || enableReadOnly;

    return (
      <div>
        <GeneralTitle>
          <h2>{t(`${sectionPath}.title`)}</h2>
        </GeneralTitle>
        <Grid container spacing={2}>
          <Grid item xs={4} sm={4}>
            <InputForm
              required
              id="airlineCode"
              label={t(`${sectionPath}.airlineIdentifier`)}
              className={
                errors.airlineCode ? 'error' : readOnly ? 'read-only-input' : ''
              }
              color="secondary"
              error={errors.airlineCode}
              defaultValue={airlineCode}
              onChange={this.handleInputChange}
              maxLength={3}
              errorMessage={t(`${COMMERCIAL_FLIGHTS_ERRORS}.airlineCode`)}
              disabled={readOnlyStatus}
            />
          </Grid>
          <Grid item xs={4} sm={4}>
            <InputForm
              required
              id="flightNumber"
              label={t(`${sectionPath}.flightNumber`)}
              className={errors.flightNumber ? 'error' : ''}
              color="secondary"
              error={errors.flightNumber}
              defaultValue={flightNumber}
              onChange={this.handleInputChange}
              maxLength={4}
              errorMessage={t(`${COMMERCIAL_FLIGHTS_ERRORS}.flightNumber`)}
              disabled={readOnlyStatus}
            />
          </Grid>
          <Grid item xs={4} sm={4}>
            <InputForm
              id="operationalSuffix"
              label={t(`${sectionPath}.operationalSuffix`)}
              className={
                errors.operationalSuffix
                  ? 'error'
                  : readOnly
                  ? 'read-only-input'
                  : ''
              }
              color="secondary"
              error={errors.operationalSuffix}
              defaultValue={operationalSuffix}
              onChange={this.handleInputChange}
              maxLength={1}
              errorMessage={t(`${COMMERCIAL_FLIGHTS_ERRORS}.operationalSuffix`)}
              disabled={readOnlyStatus}
            />
          </Grid>

          <Grid item xs={6} sm={6}>
            <Autocomplete
              name="departureStationCode"
              id="departureStationCode"
              options={prepareStations(stations)}
              value={
                departureStationCode
                  ? {
                      value: departureStationCode,
                      label: getStationName(stations, departureStationCode),
                    }
                  : null
              }
              ListboxComponent={ListboxComponent}
              onChange={(e, option) =>
                this.handleSelectChange(option, 'departureStationCode')
              }
              disabled={readOnlyStatus}
              getOptionLabel={option =>
                option.value ? `${option.value} - ${option.label}` : ''
              }
              renderInput={params => (
                <TextField
                  {...params}
                  variant="standard"
                  label={t(`${sectionPath}.departureStationCode`) + ' *'}
                />
              )}
              renderOption={option => <SingleItem>{option.label}</SingleItem>}
            />
          </Grid>
          <Grid item xs={6} sm={6}>
            <FormControl>
              <InputLabel htmlFor="passengerTerminalDeparture">
                {t(`${sectionPath}.terminal`)}
              </InputLabel>
              <SelectInput
                name="passengerTerminalDeparture"
                onChange={this.handleChange}
                items={findTerminalsByStationCode(
                  stations,
                  departureStationCode
                )}
                value={passengerTerminalDeparture}
                disabled={readOnlyStatus}
              />
            </FormControl>
          </Grid>

          <Grid item xs={6} sm={6}>
            <Autocomplete
              name="arrivalStationCode"
              id="arrivalStationCode"
              options={prepareStations(stations)}
              value={
                arrivalStationCode
                  ? {
                      value: arrivalStationCode,
                      label: getStationName(stations, arrivalStationCode),
                    }
                  : null
              }
              ListboxComponent={ListboxComponent}
              onChange={(e, option) =>
                this.handleSelectChange(option, 'arrivalStationCode')
              }
              disabled={readOnlyStatus}
              getOptionLabel={option =>
                option.value ? `${option.value} - ${option.label}` : ''
              }
              renderInput={params => (
                <TextField
                  {...params}
                  variant="standard"
                  label={t(`${sectionPath}.arrivalStationCode`) + ' *'}
                />
              )}
              renderOption={option => <SingleItem>{option.label}</SingleItem>}
            />
          </Grid>
          <Grid item xs={6} sm={6}>
            <FormControl>
              <InputLabel htmlFor="passengerTerminalArrival">
                {t(`${sectionPath}.terminal`)}
              </InputLabel>
              <SelectInput
                name="passengerTerminalArrival"
                onChange={this.handleChange}
                items={findTerminalsByStationCode(stations, arrivalStationCode)}
                value={passengerTerminalArrival}
                disabled={readOnlyStatus}
              />
            </FormControl>
          </Grid>

          <TimePickers
            required
            beforeName={'startTime'}
            beforeLabel={t(`${sectionPath}.startTime`)}
            firstValue={startTime}
            afterName={'endTime'}
            afterLabel={t(`${sectionPath}.endTime`)}
            lastValue={endTime}
            onChange={this.handleChangePicker}
            disabled={readOnlyStatus}
          />

          <Grid item xs={6} sm={6}>
            <InputForm
              required
              id="aircraftType"
              label={t(`${sectionPath}.aircraftTypeCode`)}
              className={
                errors.aircraftTypeCode
                  ? 'error'
                  : readOnly
                  ? 'read-only-input'
                  : ''
              }
              color="secondary"
              error={errors.aircraftType}
              defaultValue={aircraftType}
              onChange={this.handleInputChange}
              maxLength={4}
              errorMessage={t(`${COMMERCIAL_FLIGHTS_ERRORS}.aircraftTypeCode`)}
              disabled={readOnlyStatus}
            />
          </Grid>

          <Grid item xs={6} sm={6}>
            <InputForm
              id="aircraftConfigurationVersion"
              label={t(`${sectionPath}.aircraftConfigurationVersion`)}
              className={errors.aircraftConfigurationVersion ? 'error' : ''}
              color="secondary"
              error={errors.aircraftConfigurationVersion}
              defaultValue={aircraftConfigurationVersion}
              onChange={this.handleInputChange}
              maxLength={20}
              errorMessage={t(
                `${COMMERCIAL_FLIGHTS_ERRORS}.aircraftConfigurationVersion`
              )}
              disabled={readOnlyStatus}
            />
          </Grid>
          <Grid item xs={12} sm={12}>
            <CustomText style={{ marginBottom: 10 }}>
              {t(`${sectionPath}.flightDates`)}
            </CustomText>
            <Calendar
              onCalendarUpdates={this.handleCalendarUpdates}
              flightDates={startDates}
              startDate={this.startDate}
              endDate={this.endDate}
              carryIn={CARRY_IN_DAYS}
              carryOut={CARRY_OUT_DAYS}
              t={t}
              disabled={readOnlyStatus}
            />
          </Grid>
        </Grid>
      </div>
    );
  };

  getInformationSection = () => {
    const { t, tagsSource, readOnly, enableReadOnly } = this.props;
    const { tags } = this.state;
    const sectionPath = COMMERCIAL_FLIGHTS_INFO;
    const readOnlyStatus = readOnly || enableReadOnly;
    return (
      <React.Fragment>
        <h2>{t(`${sectionPath}.title`)}</h2>
        <Grid container spacing={2} style={{ marginBottom: '30px' }}>
          <Grid item xs={12} sm={12}>
            <CustomText id="tags" style={{ marginBottom: 10 }}>
              {t(`${sectionPath}.tags`)}
            </CustomText>
            <AutocompleteChips
              chips={tags}
              chipsSource={tagsSource}
              onChipsUpdate={this.handleTagsUpdate}
              disabled={readOnlyStatus}
              label={t(`${sectionPath}.label`)}
            />
          </Grid>
        </Grid>
      </React.Fragment>
    );
  };

  render() {
    const { selectedItem, t, readOnly, enableReadOnly, ...rest } = this.props;

    const readOnlyStatus = readOnly || enableReadOnly;
    const isDisabled = readOnlyStatus ? false : this.isReadyToSubmit();

    return (
      <Form
        okButton={getFormOkButton(t, selectedItem, readOnly)}
        cancelButton={getFormCancelButton(t, readOnlyStatus)}
        isDisabled={isDisabled}
        onChange={() => {}}
        anchor={'right'}
        enableReadOnly={enableReadOnly}
        {...rest}
        handleOk={this.handleSubmit}
      >
        <FormHeader>
          <span>
            {getFormHeading(
              t,
              selectedItem,
              readOnlyStatus,
              COMMERCIAL_FLIGHTS_FORM
            )}
          </span>
          <span>{name}</span>
        </FormHeader>
        <FormBody>
          <PerfectScrollbar option={perfectScrollConfig}>
            <div>
              {this.getGeneralSection()}
              {this.getInformationSection()}
            </div>
          </PerfectScrollbar>
        </FormBody>
      </Form>
    );
  }
}

CommercialFlightsForm.propTypes = {
  t: PropTypes.func.isRequired,
  readOnly: PropTypes.bool.isRequired,
  openItemId: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  handleOk: PropTypes.func.isRequired,
  formId: PropTypes.string.isRequired,
  handleCrewComplements: PropTypes.func.isRequired,
  selectedItem: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    name: PropTypes.string,
  }),
  isOpen: PropTypes.bool.isRequired,
  stations: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      name: PropTypes.string,
      code: PropTypes.string,
    })
  ),
  positions: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      name: PropTypes.string,
      code: PropTypes.string,
    })
  ),
  airlines: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      code: PropTypes.string,
    })
  ),
  services: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      code: PropTypes.string,
    })
  ),
  aircrafts: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      code: PropTypes.string,
    })
  ),
  tagsSource: PropTypes.arrayOf(PropTypes.string),
  enableReadOnly: PropTypes.bool,
};

CommercialFlightsForm.defaultProps = {
  selectedItem: null,
  stations: [],
  positions: [],
  airlines: [],
  services: [],
  aircrafts: [],
  tagsSource: [],
  enableReadOnly: false,
};

export default CommercialFlightsForm;
