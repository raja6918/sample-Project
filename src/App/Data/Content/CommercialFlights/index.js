import CommercialFlightsWithErrorHandling from './CommercialFlights';

export default CommercialFlightsWithErrorHandling;
