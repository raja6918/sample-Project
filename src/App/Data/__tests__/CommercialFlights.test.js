import React from 'react';
import { shallow } from 'enzyme';

import CommercialFlights from '../Content/CommercialFlights';

const t = jest.fn();

test('CommercialFlights component render correct', () => {
  const wrapper = shallow(<CommercialFlights t={t} />);

  expect(wrapper).toMatchSnapshot();
});
