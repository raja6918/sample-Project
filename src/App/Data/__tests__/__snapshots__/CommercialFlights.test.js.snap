// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommercialFlights component render correct 1`] = `
<Fragment>
  <withDeleteHandler
    clearErrorNotification={[Function]}
    inlineError={null}
    reportError={[Function]}
    t={[MockFunction]}
  />
</Fragment>
`;
